crates/rust-threadpool/src/lib.rs

```diff
@@ -0,0 +1,1378 @@
+    /// use std::thread;
+    /// use std::sync::mpsc::channel;
+    ///
+    /// let pool = ThreadPool::with_name("clone example".into(), 2);
+    ///
+    /// let results = (0..2)
+    ///     .map(|i| {
+    ///         let pool = pool.clone();
+    ///         thread::spawn(move || {
+    ///             let (tx, rx) = channel();
+    ///             for i in 1..12 {
+    ///                 let tx = tx.clone();
+    ///                 pool.execute(move || {
+    ///                     tx.send(i).expect("channel will be waiting");
+    ///                 });
+    ///             }
+    ///             drop(tx);
+    ///             if i == 0 {
+    ///                 rx.iter().fold(0, |accumulator, element| accumulator + element)
+    ///             } else {
+    ///                 rx.iter().fold(1, |accumulator, element| accumulator * element)
+    ///             }
+    ///         })
+    ///     })
+    ///     .map(|join_handle| join_handle.join().expect("collect results from threads"))
+    ///     .collect::<Vec<usize>>();
+    ///
+    /// assert_eq!(vec![66, 39916800], results);
+    /// ```
+    fn clone(&self) -> ThreadPool {
+        ThreadPool {
+            jobs: self.jobs.clone(),
+            shared_data: self.shared_data.clone(),
+        }
+    }
+}
+
+/// Create a thread pool with one thread per CPU.
+/// On machines with hyperthreading,
+/// this will create one thread per hyperthread.
+impl Default for ThreadPool {
+    fn default() -> Self {
+        ThreadPool::new(NUM_CPUS)
+    }
+}
+
+impl fmt::Debug for ThreadPool {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ThreadPool")
+            .field("name", &self.shared_data.name)
+            .field("queued_count", &self.queued_count())
+            .field("active_count", &self.active_count())
+            .field("max_count", &self.max_count())
+            .finish()
+    }
+}
+
+impl PartialEq for ThreadPool {
+    /// Check if you are working with the same pool
+    ///
+    /// ```
+    /// use threadpool::ThreadPool;
+    ///
+    /// let a = ThreadPool::new(2);
+    /// let b = ThreadPool::new(2);
+    ///
+    /// assert_eq!(a, a);
+    /// assert_eq!(b, b);
+    ///
+    /// # // TODO: change this to assert_ne in the future
+    /// assert!(a != b);
+    /// assert!(b != a);
+    /// ```
+    fn eq(&self, other: &ThreadPool) -> bool {
+        let a: &ThreadPoolSharedData = &*self.shared_data;
+        let b: &ThreadPoolSharedData = &*other.shared_data;
+        a as *const ThreadPoolSharedData == b as *const ThreadPoolSharedData
+        // with rust 1.17 and late:
+        // Arc::ptr_eq(&self.shared_data, &other.shared_data)
+    }
+}
+impl Eq for ThreadPool {}
+
+fn spawn_in_pool(shared_data: Arc<ThreadPoolSharedData>) {
+    let mut builder = thread::Builder::new();
+    if let Some(ref name) = shared_data.name {
+        builder = builder.name(name.clone());
+    }
+    // Yu Ding:
```

Comment:
Since we want to build up a open community, please it change to `XXX:`.

Reply:
Or, `Note:`

Reply:
Updated @mssun